File: SelfHomeApp/app/src/main/java/it/dani/selfhomeapp/middle/icontroller/IGroupController.java
package it.dani.selfhomeapp.middle.icontroller;

import androidx.annotation.NonNull;

import java.util.Set;
import java.util.function.Consumer;

public interface IGroupController {
    public void getGroups(@NonNull Consumer<Set<String>> consumer);
    public void setGroupState(@NonNull String groupName, int state, @NonNull Consumer<Boolean> consumer);
    public void addNewGroup(@NonNull String groupName, @NonNull Consumer<Boolean> consumer);
    public void delGroup(@NonNull String groupName, @NonNull Consumer<Boolean> consumer);
    public void getDevicesFromGroup(@NonNull String groupName, @NonNull Consumer<Set<String>> consumer);
    public void addDeviceToGroup(@NonNull String deviceName, @NonNull String groupName, @NonNull Consumer<Boolean> consumer);
    public void delDeviceToGroup(@NonNull String deviceName, @NonNull String groupName, @NonNull Consumer<Boolean> consumer);

    public void getGroups(@NonNull Consumer<Set<String>> consumer, Consumer<Exception> error);
    public void setGroupState(@NonNull String groupName, int state, @NonNull Consumer<Boolean> consumer, Consumer<Exception> error);
    public void addNewGroup(@NonNull String groupName, @NonNull Consumer<Boolean> consumer, Consumer<Exception> error);
    public void delGroup(@NonNull String groupName, @NonNull Consumer<Boolean> consumer, Consumer<Exception> error);
    public void getDevicesFromGroup(@NonNull String groupName, @NonNull Consumer<Set<String>> consumer, Consumer<Exception> error);
    public void addDeviceToGroup(@NonNull String deviceName, @NonNull String groupName, @NonNull Consumer<Boolean> consumer, Consumer<Exception> error);
    public void delDeviceToGroup(@NonNull String deviceName, @NonNull String groupName, @NonNull Consumer<Boolean> consumer, Consumer<Exception> error);
}

File: SelfHomeApp/app/src/main/java/it/dani/selfhomeapp/middle/icontroller/IGroupController.kt
package it.dani.selfhomeapp.middle.icontroller

interface IGroupController {
    fun getGroups(consumer: (Set<String>) -> Unit)
    fun setGroupState(groupName: String, state: Int, consumer: (Boolean) -> Unit)
    fun addNewGroup(groupName: String, consumer: (Boolean) -> Unit)
    fun delGroup(groupName: String, consumer: (Boolean) -> Unit)
    fun getDevicesFromGroup(groupName: String, consumer: (Set<String>) -> Unit)
    fun addDeviceToGroup(deviceName: String, groupName: String, consumer: (Boolean) -> Unit)
    fun delDeviceToGroup(deviceName: String, groupName: String, consumer: (Boolean) -> Unit)

    fun getGroups(consumer: (Set<String>) -> Unit, error: (Exception) -> Unit)
    fun setGroupState(groupName: String, state: Int, consumer: (Boolean) -> Unit, error: (Exception) -> Unit)

    fun addNewGroup(groupName: String, consumer: (Boolean) -> Unit, error: (Exception) -> Unit)
    fun delGroup(groupName: String, consumer: (Boolean) -> Unit, error: (Exception) -> Unit)
    fun getDevicesFromGroup(groupName: String, consumer: (Set<String>) -> Unit, error: (Exception) -> Unit)
    fun addDeviceToGroup(deviceName: String, groupName: String, consumer: (Boolean) -> Unit, error: (Exception) -> Unit)
    fun delDeviceToGroup(deviceName: String, groupName: String, consumer: (Boolean) -> Unit, error: (Exception) -> Unit)
}